phpBB/phpbb/install/helper/file_updater/file_updater.php

```diff
@@ -0,0 +1,174 @@
+	{
+		$this->filesystem		= $filesystem;
+		$this->phpbb_root_path	= $phpbb_root_path;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 *
+	 * @throws file_updater_failure_exception	When the file is not writable
+	 * @throws filesystem_exception				When the filesystem class fails
+	 */
+	public function delete_file($path_to_file)
+	{
+		$this->filesystem->remove($this->phpbb_root_path . $path_to_file);
+	}
+
+	/**
+	 * {@inheritdoc}
+	 *
+	 * @throws file_updater_failure_exception	When the file is not writable
+	 * @throws filesystem_exception				When the filesystem class fails
+	 */
+	public function create_new_file($path_to_file_to_create, $source, $create_from_content = false)
+	{
+		$path_to_file_to_create = $this->phpbb_root_path . $path_to_file_to_create;
+
+		$dir = $path_to_file_to_create . '../';
+		$original_dir_perms = false;
+
+		if (!$this->filesystem->is_writable($dir))
+		{
+			// Extract last 9 bits we actually need
+			$original_dir_perms = fileperms($dir) & 511;
+			$this->filesystem->phpbb_chmod($dir, filesystem::CHMOD_ALL);
+		}
+
+		if (!$create_from_content)
+		{
+			try
+			{
+				$this->filesystem->copy($source, $path_to_file_to_create);
+			}
+			catch (filesystem_exception $e)
+			{
+				$this->write_file($path_to_file_to_create, $source, $create_from_content);
+			}
+		}
+		else
+		{
+			$this->write_file($path_to_file_to_create, $source, $create_from_content);
+		}
+
+		if ($original_dir_perms !== false)
+		{
+			$this->filesystem->phpbb_chmod($dir, $original_dir_perms);
+		}
+	}
+
+	/**
+	 * {@inheritdoc}
+	 *
+	 * @throws file_updater_failure_exception	When the file is not writable
+	 * @throws filesystem_exception				When the filesystem class fails
+	 */
+	public function update_file($path_to_file_to_update, $source, $create_from_content = false)
+	{
+		$path_to_file_to_update = $this->phpbb_root_path . $path_to_file_to_update;
+		$original_file_perms = false;
+
+		if (!$this->filesystem->is_writable($path_to_file_to_update))
+		{
+			// Extract last 9 bits we actually need
+			$original_file_perms = fileperms($path_to_file_to_update) & 511;
+			$this->filesystem->phpbb_chmod($path_to_file_to_update, filesystem::CHMOD_WRITE);
+		}
+
+		if (!$create_from_content)
+		{
+			try
+			{
+				$this->filesystem->copy($source, $path_to_file_to_update, true);
+			}
+			catch (filesystem_exception $e)
+			{
+				$this->write_file($path_to_file_to_update, $source, $create_from_content);
+			}
+		}
+		else
+		{
+			$this->write_file($path_to_file_to_update, $source, $create_from_content);
+		}
+
+		if ($original_file_perms !== false)
+		{
+			$this->filesystem->phpbb_chmod($path_to_file_to_update, $original_file_perms);
+		}
+	}
+
+	/**
+	 * Fallback function for file writing
+	 *
+	 * @param string		$path_to_file			Path to the file's location
+	 * @param string		$source					Path to file to copy or string with the new file's content
+	 * @param bool|false	$create_from_content	Whether or not to use $source as the content, false by default
```

Comment:
All of them should be tabs, so i guess that is on github (also the code and discussion view is rendered differently most places, so i guess we should be fine)